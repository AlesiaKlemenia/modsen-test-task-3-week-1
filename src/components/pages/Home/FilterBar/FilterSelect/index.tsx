import { MenuItem, SelectChangeEvent } from '@mui/material';
import { FilterWrapper, SelectLabel, StyledSelect } from './styled';

export interface ISelectItem {
  name: string;
  value: string;
}

interface IFilterSelectProps {
  label: string;
  items: ISelectItem[];
  value: string;
  onSelectChange: (e: SelectChangeEvent<unknown>) => void;
}

const FilterSelect = ({
  label,
  items,
  value,
  onSelectChange,
}: IFilterSelectProps): JSX.Element => (
  <FilterWrapper>
    <SelectLabel id={`${label}-select-label}`} sx={{ color: 'rgba(0, 0, 0)' }}>
      {label}
    </SelectLabel>
    <StyledSelect
      labelId={`${label}-select-label`}
      id={`${label}-select`}
      value={value}
      onChange={onSelectChange}
    >
      {items.map((item, i) => (
        <MenuItem key={i} value={item.value}>
          {item.name}
        </MenuItem>
      ))}
    </StyledSelect>
  </FilterWrapper>
);

export default FilterSelect;
