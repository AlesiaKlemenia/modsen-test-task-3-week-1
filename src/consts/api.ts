const apiUrl = 'https://www.googleapis.com/books/v1/';

export const searchBooksUrl = `${apiUrl}volumes?q=`;
export const getBookUrl = `${apiUrl}volumes/`;
