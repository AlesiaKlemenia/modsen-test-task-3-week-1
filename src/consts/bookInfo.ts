export interface IBookCardInfo {
  id: string;
  title: string;
  bookCategories: string[];
  authors: string[];
  coverUrl: string;
}

export interface IBookDetailInfo extends IBookCardInfo {
  description: string;
}

export interface IFullBookInfo {
  accessInfo: object;
  etag: string;
  id: string;
  kind: string;
  saleInfo: object;
  searchInfo: object;
  volumeInfo: {
    allowAnonLogging: boolean;
    authors: string[];
    categories: string[];
    contentVersion: string;
    description: string;
    imageLinks: {
      smallThumbnail: string;
      thumbnail: string;
    };
    industryIdentifiers: [];
    infoLink: string;
    language: string;
    maturityRating: string;
    pageCount: number;
    panelizationSummary: object;
    previewLink: string;
    printType: string;
    publishedDate: string;
    readingModes: object;
    title: string;
  };
}
