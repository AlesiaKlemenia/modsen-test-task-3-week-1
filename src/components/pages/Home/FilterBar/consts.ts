import { ISelectItem } from './FilterSelect';

export const categoryItems: ISelectItem[] = [
  {
    name: 'All',
    value: 'all',
  },
  {
    name: 'Art',
    value: 'art',
  },
  {
    name: 'Biography',
    value: 'biography',
  },
  {
    name: 'Computers',
    value: 'computers',
  },
  {
    name: 'History',
    value: 'history',
  },
  {
    name: 'Medicial',
    value: 'medicial',
  },
  {
    name: 'Poetry',
    value: 'poetry',
  },
];

export const sortingItems: ISelectItem[] = [
  {
    name: 'Relevance',
    value: 'relevance',
  },
  {
    name: 'Newest',
    value: 'newest',
  },
];
