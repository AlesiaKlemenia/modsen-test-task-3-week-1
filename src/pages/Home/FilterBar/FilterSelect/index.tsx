import { InputLabel, MenuItem } from '@mui/material';

import { IFilterSelectProps } from '@/pages/Home/FilterBar/FilterSelect/interfaces';
import {
  FilterWrapper,
  StyledSelect,
} from '@/pages/Home/FilterBar/FilterSelect/styled';

const FilterSelect = ({
  label,
  items,
  value,
  onSelectChange,
}: IFilterSelectProps): JSX.Element => (
  <FilterWrapper>
    <InputLabel id={`${label}-select-label}`} sx={{ color: 'rgba(0, 0, 0)' }}>
      {label}
    </InputLabel>
    <StyledSelect
      labelId={`${label}-select-label`}
      id={`${label}-select`}
      value={value}
      onChange={onSelectChange}
    >
      {items.map((item, i) => (
        <MenuItem key={i} value={item.value}>
          {item.name}
        </MenuItem>
      ))}
    </StyledSelect>
  </FilterWrapper>
);

export default FilterSelect;
