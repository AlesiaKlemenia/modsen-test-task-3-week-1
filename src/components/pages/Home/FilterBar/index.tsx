import { IFilterBarProps } from '../../../../interfaces/IFilterBarProps';
import { categoryItems, sortingItems } from './consts';
import FilterSelect from './FilterSelect';
import { FilterBarWrapper } from './styled';

const FilterBar = ({
  category,
  onCategorySelectChange,
  sort,
  onSortingSelectChange,
}: IFilterBarProps): JSX.Element => (
  <FilterBarWrapper>
    <FilterSelect
      label="Category"
      items={categoryItems}
      value={category}
      onSelectChange={onCategorySelectChange}
    />
    <FilterSelect
      label="Sorting"
      items={sortingItems}
      value={sort}
      onSelectChange={onSortingSelectChange}
    />
  </FilterBarWrapper>
);

export default FilterBar;
