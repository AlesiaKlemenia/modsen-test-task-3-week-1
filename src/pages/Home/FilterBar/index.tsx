import { IFilterBarProps } from '@/interfaces/IFilterBarProps';
import { categoryItems, sortingItems } from '@/pages/Home/FilterBar/consts';
import FilterSelect from '@/pages/Home/FilterBar/FilterSelect';
import { FilterBarWrapper } from '@/pages/Home/FilterBar/styled';

const FilterBar = ({
  category,
  onCategorySelectChange,
  sort,
  onSortingSelectChange,
}: IFilterBarProps): JSX.Element => (
  <FilterBarWrapper>
    <FilterSelect
      label="Category"
      items={categoryItems}
      value={category}
      onSelectChange={onCategorySelectChange}
    />
    <FilterSelect
      label="Sorting"
      items={sortingItems}
      value={sort}
      onSelectChange={onSortingSelectChange}
    />
  </FilterBarWrapper>
);

export default FilterBar;
