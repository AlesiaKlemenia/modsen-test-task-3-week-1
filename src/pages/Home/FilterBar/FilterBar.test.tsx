import '@testing-library/jest-dom';

import { StyledEngineProvider } from '@mui/material';
import renderer from 'react-test-renderer';

import FilterBar from '@/pages/Home/FilterBar';
import { categoryItems, sortingItems } from '@/pages/Home/FilterBar/consts';

const mockCategorySelectChange = jest.fn();
const mockSortingSelectChange = jest.fn();

describe('Filter Bar component', () => {
  it('renders correctly with correct props', () => {
    const component = renderer.create(
      <StyledEngineProvider injectFirst>
        <FilterBar
          category={categoryItems[0].value}
          onCategorySelectChange={mockCategorySelectChange}
          sort={sortingItems[0].value}
          onSortingSelectChange={mockSortingSelectChange}
        />
      </StyledEngineProvider>,
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('calls onCategorySelectChange when category is selected', () => {
    const component = renderer.create(
      <StyledEngineProvider injectFirst>
        <FilterBar
          category={categoryItems[0].value}
          onCategorySelectChange={mockCategorySelectChange}
          sort={sortingItems[0].value}
          onSortingSelectChange={mockSortingSelectChange}
        />
      </StyledEngineProvider>,
    );

    const instance = component.root;

    const categorySelect = instance.findByProps({ label: 'Category' });
    categorySelect.props.onSelectChange(categoryItems[1].value);
    expect(mockCategorySelectChange).toHaveBeenCalledWith(
      categoryItems[1].value,
    );
  });

  it('calls onSortingSelectChange when sorting option is selected', () => {
    const component = renderer.create(
      <StyledEngineProvider injectFirst>
        <FilterBar
          category={categoryItems[0].value}
          onCategorySelectChange={mockCategorySelectChange}
          sort={sortingItems[0].value}
          onSortingSelectChange={mockSortingSelectChange}
        />
      </StyledEngineProvider>,
    );

    const instance = component.root;

    const sortingSelect = instance.findByProps({ label: 'Sorting' });
    sortingSelect.props.onSelectChange(sortingItems[1].value);
    expect(mockSortingSelectChange).toHaveBeenCalledWith(sortingItems[1].value);
  });

  it('renders correct number of category items', () => {
    const component = renderer.create(
      <StyledEngineProvider injectFirst>
        <FilterBar
          category={categoryItems[0].value}
          onCategorySelectChange={mockCategorySelectChange}
          sort={sortingItems[0].value}
          onSortingSelectChange={mockSortingSelectChange}
        />
      </StyledEngineProvider>,
    );

    const instance = component.root;
    const categorySelect = instance.findByProps({ label: 'Category' });

    expect(categorySelect.props.items).toHaveLength(categoryItems.length);
  });

  it('renders correct number of sorting items', () => {
    const component = renderer.create(
      <StyledEngineProvider injectFirst>
        <FilterBar
          category={categoryItems[0].value}
          onCategorySelectChange={mockCategorySelectChange}
          sort={sortingItems[0].value}
          onSortingSelectChange={mockSortingSelectChange}
        />
      </StyledEngineProvider>,
    );

    const instance = component.root;
    const sortingSelect = instance.findByProps({ label: 'Sorting' });

    expect(sortingSelect.props.items).toHaveLength(sortingItems.length);
  });

  it('renders correct category value', () => {
    const component = renderer.create(
      <StyledEngineProvider injectFirst>
        <FilterBar
          category={categoryItems[1].value}
          onCategorySelectChange={mockCategorySelectChange}
          sort={sortingItems[0].value}
          onSortingSelectChange={mockSortingSelectChange}
        />
      </StyledEngineProvider>,
    );

    const instance = component.root;
    const categorySelect = instance.findByProps({ label: 'Category' });

    expect(categorySelect.props.value).toBe(categoryItems[1].value);
  });

  it('renders correct sorting value', () => {
    const component = renderer.create(
      <StyledEngineProvider injectFirst>
        <FilterBar
          category={categoryItems[0].value}
          onCategorySelectChange={mockCategorySelectChange}
          sort={sortingItems[1].value}
          onSortingSelectChange={mockSortingSelectChange}
        />
      </StyledEngineProvider>,
    );

    const instance = component.root;
    const sortingSelect = instance.findByProps({ label: 'Sorting' });

    expect(sortingSelect.props.value).toBe(sortingItems[1].value);
  });
});
