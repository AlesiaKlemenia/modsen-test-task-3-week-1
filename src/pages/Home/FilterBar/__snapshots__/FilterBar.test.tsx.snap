// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter Bar component renders correctly with correct props 1`] = `
<div
  className="sc-hLQSwg btxuYP"
>
  <div
    className="sc-dmyCSP jyidcN"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-animated MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-animated sc-beySPh kylVPf css-9npbnl-MuiFormLabel-root-MuiInputLabel-root"
      id="Category-select-label}"
    >
      Category
    </label>
    <div
      className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary sc-guDLey qrKYn css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
      onClick={[Function]}
    >
      <div
        aria-controls=":r0:"
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-labelledby="Category-select-label Category-select"
        className="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
        id="Category-select"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        role="combobox"
        tabIndex={0}
      >
        All
      </div>
      <input
        aria-hidden={true}
        aria-invalid={false}
        className="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
        onAnimationStart={[Function]}
        onChange={[Function]}
        tabIndex={-1}
        value="all"
      />
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
        data-testid="ArrowDropDownIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
      <fieldset
        aria-hidden={true}
        className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          className="css-ihdtdm"
        >
          <span
            className="notranslate"
          >
            ​
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
  <div
    className="sc-dmyCSP jyidcN"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-animated MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-animated sc-beySPh kylVPf css-9npbnl-MuiFormLabel-root-MuiInputLabel-root"
      id="Sorting-select-label}"
    >
      Sorting
    </label>
    <div
      className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary sc-guDLey qrKYn css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
      onClick={[Function]}
    >
      <div
        aria-controls=":r1:"
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-labelledby="Sorting-select-label Sorting-select"
        className="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
        id="Sorting-select"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        role="combobox"
        tabIndex={0}
      >
        Relevance
      </div>
      <input
        aria-hidden={true}
        aria-invalid={false}
        className="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
        onAnimationStart={[Function]}
        onChange={[Function]}
        tabIndex={-1}
        value="relevance"
      />
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
        data-testid="ArrowDropDownIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
      <fieldset
        aria-hidden={true}
        className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          className="css-ihdtdm"
        >
          <span
            className="notranslate"
          >
            ​
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;
